types/src/v0/impls/stake_table.rs

```diff
@@ -239,168 +215,220 @@ pub fn validators_from_l1_events<I: Iterator<Item = StakeTableEvent>>(
+                entry.or_insert(Validator {
+                    account,
+                    stake_table_key,
+                    state_ver_key,
+                    stake: U256::ZERO,
+                    commission,
+                    delegators: HashMap::new(),
+                });
 
-                match validators.entry(account) {
-                    indexmap::map::Entry::Occupied(_occupied_entry) => {
-                        bail!("validator {:#x} already registered", *account)
-                    },
-                    indexmap::map::Entry::Vacant(vacant_entry) => vacant_entry.insert(Validator {
-                        account,
-                        stake_table_key,
-                        state_ver_key,
-                        stake: U256::from(0_u64),
-                        commission,
-                        delegators: HashMap::default(),
-                    }),
-                };
+                Ok(())
             },
-            StakeTableEvent::RegisterV2(event) => {
+
+            StakeTableEvent::RegisterV2(reg) => {
                 // Signature authentication is performed right after fetching, if we get an
                 // unauthenticated event here, something went wrong, we abort early.
-                event
-                    .authenticate()
-                    .context("Failed to authenticate event: {event:?}")?;
+                reg.authenticate()
+                    .map_err(|e| StakeTableError::AuthenticationFailed(e.to_string()))?;
+
                 let ValidatorRegisteredV2 {
                     account,
                     blsVK,
                     schnorrVK,
                     commission,
                     ..
-                } = event;
+                } = reg;
 
                 let stake_table_key: BLSPubKey = blsVK.into();
                 let state_ver_key: SchnorrPubKey = schnorrVK.into();
+
                 // The stake table contract enforces that each bls key is only used once.
-                if bls_keys.contains(&stake_table_key) {
-                    bail!("bls key already used: {}", stake_table_key.to_string());
-                };
+                if !self.used_bls_keys.insert(stake_table_key) {
+                    tracing::warn!(
+                        stake_table_key = %stake_table_key,
+                        "Duplicate BLS key found during stake table events processing"
+                    );
+                    return Err(StakeTableError::BlsKeyAlreadyUsed(
+                        stake_table_key.to_string(),
+                    ));
+                }
 
                 // The contract does *not* enforce that each schnorr key is only used once.
-                if schnorr_keys.contains(&state_ver_key) {
-                    tracing::warn!("schnorr key already used: {}", state_ver_key.to_string());
-                };
+                if !self.used_schnorr_keys.insert(state_ver_key.clone()) {
```

Comment:
Maybe we should just enforce these keys to be unique in the contract too. I think we discussed that it makes sense but we're also thinking about totally rewriting the stake table. :thinking: let's discuss outside this PR.